### PR TITLE
Add test for host registration through LB

### PR DESCRIPTION
**Pull request by shweta83**

### CODE REVIEW

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 1]
```diff
@@ -227,3 +227,47 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
...
+    :id: c7e47d61-167b-4fc2-8d1a-d9a64350fdc4
+
+    :Steps:
+        1. run `subscription-manager register --org=Your_Organization \
+            --activationkey=Your_Activation_Key \
+            --serverurl=https://loadbalancer.example.com:8443/rhsm \
+            --baseurl=https://loadbalancer.example.com/pulp/content`
+        2. Register host through loadbalancer using global registration
+        3. Try package installation
+    :expectedresults: Global Registration should have option to register through
```
Comment (adarshdubey-star) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:247]:
```suggestion

    :expectedresults: Global Registration should have option to register through
```

**Review by adarshdubey-star — APPROVED** [review]
Minor change requested rest looks good.

**Review by Griffin-Sullivan — APPROVED** [review]
Looks good just a couple questions

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 2]
```diff
@@ -142,7 +142,7 @@ def loadbalancer_setup(
 def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, rhel7_contenthost):
     """Register the client using ak to the haproxy
 
-    :id: bd3c2e50-18e2-4be7-8a7f-c32472e17c61
+    :id: 2125f9f6-62a4-4d84-910c-57737a617cd7
```
Comment (shweta83) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:145]: This is updated as part of pre-commit check.

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 3]
```diff
@@ -227,3 +227,47 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
...
+
+    :id: c7e47d61-167b-4fc2-8d1a-d9a64350fdc4
+
+    :Steps:
+        1. run `subscription-manager register --org=Your_Organization \
+            --activationkey=Your_Activation_Key \
+            --serverurl=https://loadbalancer.example.com:8443/rhsm \
+            --baseurl=https://loadbalancer.example.com/pulp/content`
+        2. Register host through loadbalancer using global registration
+        3. Try package installation
```
[the remark this answers is not in the document]
Comment (shweta83) [reply tests/foreman/destructive/test_capsule_loadbalancer.py]: Done.

File: tests/foreman/destructive/test_capsule_loadbalancer.py (hunk 2 above)
Comment (Gauravtalreja1) [reply tests/foreman/destructive/test_capsule_loadbalancer.py:145]: can you check if there's any test with same uuid and revert this? non-blocking

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 4]
```diff
@@ -227,3 +227,49 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
         {'organization-id': loadbalancer_setup['module_org'].id}
     )
     assert rhel7_contenthost.hostname in [host['name'] for host in hosts]
+
+
+@pytest.mark.tier1
+def test_client_register_through_lb(
+    loadbalancer_setup, rhel7_contenthost, module_target_sat, module_org, module_ak, module_location
```
Comment (Gauravtalreja1) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:234]: can we use `rhel_contenthost` fixture here?

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 5]
```diff
@@ -227,3 +227,49 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
...
+):
+    """Register the client through loadbalancer
+
+    :id: c7e47d61-167b-4fc2-8d1a-d9a64350fdc4
+
+    :Steps:
+
```
Comment (Gauravtalreja1) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:241]:
```suggestion
```

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 6]
```diff
@@ -227,3 +227,49 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
...
+    :CaseLevel: Integration
+    """
+    url = f'https://{loadbalancer_setup["setup_haproxy"]["haproxy"].hostname}'
+    server_url = f'{url}:8443/rhsm'
+    base_url = f'{url}/pulp/content'
+
+    result = rhel7_contenthost.download_install_rpm(
+        repo_url=f'{url}/pub', package_name='katello-ca-consumer-latest.noarch'
+    )
+    assert result.status == 0
```
Comment (Gauravtalreja1) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:261]: katello-ca-consumer rpm will be installed by global registration, so it looks redundant

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 7]
```diff
@@ -227,3 +227,49 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
...
+
+    command = module_target_sat.api.RegistrationCommand(
+        org=loadbalancer_setup['module_org'].label,
+        activation_key=loadbalancer_setup['content_for_client']['client_ak'].name,
+        serverurl=server_url,
+        baseurl=base_url,
```
Comment (Gauravtalreja1) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:267]: RegistrationCommand entity doesn't contain any such field in nailgun, could you check please?

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 8]
```diff
@@ -227,3 +227,49 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
...
+    ).create()
+    result = rhel7_contenthost.execute(command)
+    assert result.status == 0, f'Failed to register host: {result.stderr}'
```
Comment (Gauravtalreja1) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:270]: can we add some more assertations here for validating subscription_config like hostname or ports?

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 9]
```diff
@@ -227,3 +227,49 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
...
+
+    :id: c7e47d61-167b-4fc2-8d1a-d9a64350fdc4
+
+    :Steps:
+
+        1. Try package installation
+        2. run `subscription-manager register --org=Your_Organization \
+            --activationkey=Your_Activation_Key \
+            --serverurl=https://loadbalancer.example.com:8443/rhsm \
+            --baseurl=https://loadbalancer.example.com/pulp/content`
```
Comment (Gauravtalreja1) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:246]: I think we're not manually registering here, so could you update the steps here?

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 10]
```diff
@@ -227,3 +227,47 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
         {'organization-id': loadbalancer_setup['module_org'].id}
     )
     assert rhel7_contenthost.hostname in [host['name'] for host in hosts]
+
+
+@pytest.mark.tier1
+def test_client_register_through_lb(
```
[the remark this answers is not in the document]
Comment (jyejare) [reply tests/foreman/destructive/test_capsule_loadbalancer.py:233]: Nop. As of today, we can run this test on upgrade but its missing destructive mark which would have given an upgraded satellite to create the load balancer setup. But I wonder why we dont have it. Please think about it !

But even after that, we will only have upgrade satellite and not capsules or clients. So if we are ok with (Upgraded satellite with non upgraded capsule and clients scenario) , we should be good.

**Comment by Ron-Lavi** [comment]
Hey @shweta83, what is the status here? :)

**Comment by shweta83** [comment]
> Hey @shweta83, what is the status here? :)

Changes are done. Just testing it on local before pushing it.

**Comment by shweta83** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_capsule_loadbalancer.py
pod_resources_size: extra_extra_large

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 11]
```diff
@@ -138,6 +150,43 @@ def loadbalancer_setup(
     }
 
 
+@pytest.fixture(scope='module')
+def enable_subscriptions(module_target_sat, module_entitlement_manifest_org, version):
+    """Enable and sync RHEL rpms repos"""
```
Comment (Gauravtalreja1) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:155]: there is a fixture `content_for_client` already that syncs custom RH repos, why do we need to sync it again from CDN? Do you think it's redundant or is it intended?

**Review by Gauravtalreja1 — COMMENTED** [review]
@shweta83 Thanks for addressing previous comment, added a few more comments, and could you please check failing PRT?

**Comment by shweta83** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_capsule_loadbalancer.py
pod_resources_size: extra_extra_large

File: tests/foreman/destructive/test_capsule_loadbalancer.py [hunk 12]
```diff
@@ -227,3 +238,47 @@ def test_loadbalancer_register_client_using_ak_to_ha_proxy(loadbalancer_setup, r
...
+    """
+    result = rhel_contenthost.register(
+        module_org,
+        module_location,
+        loadbalancer_setup['content_for_client']['client_ak'].name,
+        satellite=module_target_sat,
+        force=True,
+    )
+    assert result.status == 0, f'Failed to register host: {result.stderr}'
+    assert module_target_sat.hostname in rhel_contenthost.subscription_config['server']['hostname']
```
Comment (Gauravtalreja1) [inline tests/foreman/destructive/test_capsule_loadbalancer.py:279]: shouldn't we assert if its registered to load balancer?
```suggestion
    assert loadbalancer_setup['content_for_client']['client_ak'].name in rhel_contenthost.subscription_config['server']['hostname']
```

**Comment by shweta83** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_capsule_loadbalancer.py
pod_resources_size: extra_extra_large

**Comment by shweta83** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_capsule_loadbalancer.py
pod_resources_size: extra_extra_large

**Comment by shweta83** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_capsule_loadbalancer.py -k test_client_register_through_lb
pod_resources_size: extra_extra_large

**Comment by shweta83** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_capsule_loadbalancer.py -k test_client_register_through_lb
pod_resources_size: extra_extra_large

**Comment by shweta83** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_capsule_loadbalancer.py -k test_client_register_through_lb
pod_resources_size: extra_extra_large

**Review by Gauravtalreja1 — APPROVED** [review]
Ack, pending results.

@shweta83 Thanks for being persistent, great work!

**Comment by shweta83** [comment]
trigger: test-robottelo
pytest: tests/foreman/destructive/test_capsule_loadbalancer.py -k test_client_register_through_lb
pod_resources_size: extra_extra_large

**Comment by shweta83** [comment]
2 test cases passed in PRT and 1 has failed because Jenkins is in maintenance. 
Test cases are tested locally and it is passing.

pytest tests/foreman/destructive/test_capsule_loadbalancer.py -k test_client_register_through_lb -sv

============================================================== test session starts ===============================================================
cachedir: .pytest_cache
shared_function enabled - OFF - scope: None - storage: file
plugins: ibutsu-2.2.4, xdist-3.2.1, mock-3.10.0, reportportal-5.1.6, services-2.2.1, cov-3.0.0
collecting ... 2023-04-11 14:27:43 - robottelo.collection - INFO - Processing test items to add testimony token markers
collected 4 items / 1 deselected / 3 selected                                                                                                    

tests/foreman/destructive/test_capsule_loadbalancer.py::test_client_register_through_lb[rhel7_contenthost_module0-rhel7] PASSED
tests/foreman/destructive/test_capsule_loadbalancer.py::test_client_register_through_lb[rhel7_contenthost_module0-rhel8] PASSED
tests/foreman/destructive/test_capsule_loadbalancer.py::test_client_register_through_lb[rhel7_contenthost_module0-rhel9] PASSED
================================================================ warnings summary ================================================================
tests/foreman/destructive/test_capsule_loadbalancer.py: 492 warnings

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================================== 3 passed, 1 deselected, 498 warnings in 5516.91s (1:31:56) ===========================================

**Review by Gauravtalreja1 — APPROVED** [review]

**Pull request merged**